### PR TITLE
Add CloudSQL support for "create from clone"

### PR DESCRIPTION
**Pull request by jasonvigil**
### Change description

Depends on #2393 

- [ ] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (maqiuyujoyce) [inline apis/sql/v1beta1/sqlinstance_types.go:463]: This looks like a reference to me. Can we reference a SQLInstance here?

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (maqiuyujoyce) [inline apis/sql/v1beta1/sqlinstance_types.go:456]: This can potentially be references to SQLDatabase resources, right?

**Review by maqiuyujoyce — COMMENTED** [review]
Code LGTM except for the potential reference fields. I assume you'll add test cases for this feature, right?

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (jasonvigil) [reply apis/sql/v1beta1/sqlinstance_types.go]: Good catch. Thanks. Fixed!

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (jasonvigil) [reply apis/sql/v1beta1/sqlinstance_types.go]: Fixed! Thank you

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (maqiuyujoyce) [inline apis/sql/v1beta1/sqlinstance_types.go:445]: The field description generated, right? The format is ... weird...

File: pkg/controller/direct/sql/mapping.go
Comment (maqiuyujoyce) [inline pkg/controller/direct/sql/mapping.go:34]: ~Does it mean that while the field is supported in the CRD, the feature can't be used?~

Okay, so basically this conversion is only triggered by the insert function, and the insert function should never be called if `in.Spec.CloneSource != nil`... got it.

Maybe the error message can be improved in the caller (insertInstance() function) saying something like: `error inserting a Cloud SQL instance with no cloud source: %w`, or you can make the error message more verbose here; otherwise this is an implicit knowledge which is not intuitive to future contributors.

File: pkg/controller/direct/sql/mapping.go
Comment (maqiuyujoyce) [inline pkg/controller/direct/sql/mapping.go:806]: Similarly, I wondered if the error message can be more verbose and actionable. E.g. `cannot convert nil CloneSource, CloudSource is required`.

File: pkg/controller/direct/sql/normalize.go
Comment (maqiuyujoyce) [inline pkg/controller/direct/sql/normalize.go:356]: Not a blocker,  but I believe we also want to check and ensure the list contains all external values or name values, but not a mix of both.

File: pkg/webhook/immutable_fields_validator.go
Comment (maqiuyujoyce) [inline pkg/webhook/immutable_fields_validator.go:386]: It's out of scope, though we probably need to think about how to handle use cases when a user realizes they made a mistake and want to fix it in-place after the creation/clone failed. Making the field strictly immutable means a user needs to delete the object, edit the manifest, and reapply, and could be a bit troublesome for some GitOps workflow.

File: pkg/webhook/immutable_fields_validator.go
Comment (maqiuyujoyce) [inline pkg/webhook/immutable_fields_validator.go:389]: I think we want to achieve the immutability check via CEL rules? Not a blocker though.

**Review by maqiuyujoyce — COMMENTED** [review]
Overall lgtm with some nits and open questions, i.e. none of the comment is a blocker.

Have you got the chance to test the clone locally? Do you plan to support test coverage and mockgcp coverage in this PR?

File: pkg/controller/direct/sql/normalize.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/normalize.go:356]: Why? I think it should be ok to mix

File: pkg/webhook/immutable_fields_validator.go
Comment (jasonvigil) [reply pkg/webhook/immutable_fields_validator.go:386]: I think that we should require delete + recreate, rather than build in some special other mechanism, but maybe we can wait for product feedback on this.

File: pkg/controller/direct/sql/mapping.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/mapping.go:34]: This error would only happen if developer misuses the conversion function. Probably just adding a comment would be sufficient.

File: pkg/controller/direct/sql/mapping.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/mapping.go:806]: Same here, this error would only happen if developer misuses the function. I'll add a comment.

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (jasonvigil) [reply apis/sql/v1beta1/sqlinstance_types.go:445]: Field description was taken directly from the client. Will update it to match the other comments.

File: pkg/controller/direct/sql/normalize.go
Comment (maqiuyujoyce) [reply pkg/controller/direct/sql/normalize.go:356]: @yuwenma proposed it and suggested it will help us avoid complicated edge cases. Yuwen, could you elaborate on it a bit?

Comment (jasonvigil) [reply pkg/controller/direct/sql/normalize.go:356]: I think that it is not really complicated to implement this, and customer may want the functionality. There could be some databases managed by kcc, others managed externally, but customer might want to clone all of them.

File: pkg/webhook/immutable_fields_validator.go
Comment (jasonvigil) [reply pkg/webhook/immutable_fields_validator.go:386]: I think it's more idiomatic to require create + recreate. But maybe this could be partially solved by educating customer about k8s best practices. Or, perhaps there is some k8s-native pattern here the customer had suggested that I am unaware of.

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (jasonvigil) [reply apis/sql/v1beta1/sqlinstance_types.go]: fixed

File: pkg/controller/direct/sql/mapping.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/mapping.go:37]: fixed

File: pkg/controller/direct/sql/mapping.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/mapping.go:810]: fixed

File: pkg/webhook/immutable_fields_validator.go
Comment (jasonvigil) [reply pkg/webhook/immutable_fields_validator.go]: fixed

File: pkg/webhook/immutable_fields_validator.go
Comment (jasonvigil) [reply pkg/webhook/immutable_fields_validator.go]: fixed

**Comment by jasonvigil** [comment]
> Overall lgtm with some nits and open questions, i.e. none of the comment is a blocker.
> 
> Have you got the chance to test the clone locally? Do you plan to support test coverage and mockgcp coverage in this PR?

Yes, tested locally. Added a basic mockgcp test.

File: pkg/controller/direct/sql/normalize.go
Comment (maqiuyujoyce) [reply pkg/controller/direct/sql/normalize.go:356]: SG!

Personally, I like the flexibility of supporting a mix of both. Implementation-wise, it's fine to start with a stricter rule (don't allow the mix) and then loosen it (allow the mix) if we think this provides a better UX.

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (maqiuyujoyce) [inline apis/sql/v1beta1/sqlinstance_types.go:472]: A silly question: Does the type of the field matter for the validation? E.g. is it able to understand "equals" for object type here?

File: pkg/test/resourcefixture/testdata/basic/sql/v1beta1/sqlinstance/sqlinstanceclonebasic/create.yaml
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/sql/v1beta1/sqlinstance/sqlinstanceclonebasic/create.yaml:18]: It's out of scope, so bringing it up just for discussion.

I personally like consistent naming for easier validation, maintenance, education (contributor may ask about how to name the resource), and potential automation. That's why we had this guide: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/docs/archive/README.NewResourceFromTerraform.md#naming-conventions-special-variables-and-constants-in-testdata

Now we are moving to direct controllers. Do you find any preferred way of naming the test resources? Or do you find any other purposes (other than consistency, maintenance easiness, etc) more important when coming up with the names?

File: pkg/test/resourcefixture/testdata/basic/sql/v1beta1/sqlinstance/sqlinstanceclonebasic/create.yaml
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/sql/v1beta1/sqlinstance/sqlinstanceclonebasic/create.yaml:29]: Nice!

**Review by maqiuyujoyce — COMMENTED** [review]
/lgtm
/approve

/hold
I wondered if @yuwenma wants to take another look.

Besides, can we add a sample about it? https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/README.Samples.md

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (jasonvigil) [reply apis/sql/v1beta1/sqlinstance_types.go:472]: Yes, equality works for maps: https://github.com/google/cel-spec/blob/master/doc/langdef.md#equality. I also tested this manually.

File: pkg/test/resourcefixture/testdata/basic/sql/v1beta1/sqlinstance/sqlinstanceclonebasic/create.yaml
Comment (jasonvigil) [reply pkg/test/resourcefixture/testdata/basic/sql/v1beta1/sqlinstance/sqlinstanceclonebasic/create.yaml:18]: I feel like it doesn't really matter that much, as long as the test resource names are unique and meaningful / understandable at first glance. Much like variable names.

File: pkg/controller/direct/sql/normalize.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/normalize.go:356]: I think it's more complicated to add a check to ensure that all references in the list are either external or kcc-managed. That requires more code. However, am not sure about the scenario you mentioned where it could cause "complicated edge cases". Do you remember the details?

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (justinsb) [reply apis/sql/v1beta1/sqlinstance_types.go:472]: I am a little wary of introducing CEL (1) on a high profile resource and (2) accidentally i.e. as part of a wider PR.  (2) is in case we need to revert and so that we can focus on CEL.

File: pkg/controller/direct/sql/merge.go
Comment (justinsb) [inline pkg/controller/direct/sql/merge.go:631]: We might also be able to look at the updateTime...

File: pkg/controller/direct/sql/normalize.go
Comment (justinsb) [inline pkg/controller/direct/sql/normalize.go:388]: FYI, k8s has fieldpath to help with this problem (writing annoyingly long field paths!)

File: pkg/controller/direct/sql/sqlinstance_controller.go
Comment (justinsb) [inline pkg/controller/direct/sql/sqlinstance_controller.go:171]: Nit: I personally think it reads better to use the "ing" form

i.e.

`failed to reconcile: waiting for SQLInstance %s clone: ...`

vs

`failed to reconcile: wait SQLInstance %s clone: ...`

File: pkg/controller/direct/sql/normalize.go
Comment (yuwenma) [reply pkg/controller/direct/sql/normalize.go:356]: Thanks @maqiuyujoyce for raising this. 
I put the rationale and some thoughts [here](https://docs.google.com/document/d/1gdbFRW0nfny8j7C_bUDuGG2Bk3z_WaC_eYhoLbuppG0/edit?tab=t.0#heading=h.g21v7n6bvl3z). The no-mix is a recommendation ("shall" rather than "must", see [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119#section-3)) . And that it is for GCP unique list.

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (yuwenma) [reply apis/sql/v1beta1/sqlinstance_types.go:472]: I proposed using CEL for immutable check in the [CR Validation doc](https://docs.google.com/document/d/1nKB6hkMV34K0SVtOdFyGZ48D4NhKhGAOTGXPEVLNAqQ/edit?resourcekey=0-hbpIoeozx9it1GVTZDIbBw&tab=t.0#bookmark=id.6quka07soumj) and put the reasons on how the decision is made:

- Kubernetes supports CEL since 1.26 which is the oldest GKE cluster version. Using CEL shall not cause GKE cluster version issues, but we shall reevaluate this when the condition changes.
- Some KCC resources use webhook to validate immutable fields. We can continue using that, but for external contributors CEL is much easier for self-learning. 
- We only apply CEL on immutable fields. More complicated CEL validation requires further discussion.

Regarding to Justin's concerns, +1 to split the immutable check in a separate PR. Because 1. we do want to use CEL for immutable field check and makes it as the trends. 2. users may hit the CEL-supportability problem when they install KCC in a non-GKE cluster and the k8s is very old. Though the possibility of this case itself is low, SQL is a high profile resource like Justin mentioned. So splitting the PR is a good idea, we can still use CEL and keep us with the clean rollback options.

**Comment by yuwenma** [comment]
/lgtm

Looks great! You may want to resolve the comments before merging.

File: pkg/controller/direct/sql/sqlinstance_controller.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/sqlinstance_controller.go]: Fixed, and for all the other SQLInstance error messages also.

File: pkg/controller/direct/sql/normalize.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/normalize.go:388]: Can you share a link to an example of this?

File: pkg/controller/direct/sql/merge.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/merge.go:631]: It felt like this was a more reliable way to do it, but am planning to switch to use the mapper / generator code soon anyway.

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (jasonvigil) [reply apis/sql/v1beta1/sqlinstance_types.go]: Ok, moving this CEL validation to a separate PR.

File: pkg/controller/direct/sql/merge.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/merge.go:631]: Though maybe the mapper generator code wouldn't really help here. Can you elaborate a bit on what you mean by using updateTime? Not sure I understand completely.

File: pkg/controller/direct/sql/normalize.go
Comment (jasonvigil) [reply pkg/controller/direct/sql/normalize.go:356]: Just spoke w/ @justinsb about this. We are thinking it is overkill to use refs for these SQLDatabases anyway, since they are associated with the parent SQLInstanceRef. Decided to make this field just a list of strings of database names.

File: apis/sql/v1beta1/sqlinstance_types.go
Comment (jasonvigil) [reply apis/sql/v1beta1/sqlinstance_types.go]: Actually we have gone back on this decision, and are thinking it's overkill to use refs here. Switched back to just a list of names. https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2429#discussion_r1729307605

**Review by justinsb — APPROVED** [review]

**Pull request merged**
